src/participant/streamParsing.ts

```diff
@@ -0,0 +1,95 @@
+function escapeRegex(str: string): string {
+  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
+}
+
+/**
+ * This function, provided an stream of text fragments, will stream the
+ * content to the provided stream and call the onIdentifierStreamed function
+ * when an identifier is streamed. This is useful for inserting code actions
+ * into a chat response, whenever a code block has been written.
+ */
+export async function processStreamWithInsertionsOnIdentifier({
+  processStreamFragment,
+  onIdentifierStreamed,
+  inputIterable,
+  identifier,
+}: {
+  processStreamFragment: (fragment: string) => void;
+  onIdentifierStreamed: (content: string) => void;
+  inputIterable: AsyncIterable<string>;
+  identifier: {
+    start: string;
+    end: string;
+  };
+}): Promise<void> {
+  const escapedIdentifierStart = escapeRegex(identifier.start);
+  const escapedIdentifierEnd = escapeRegex(identifier.end);
+  const regex = new RegExp(
+    `${escapedIdentifierStart}([\\s\\S]*?)${escapedIdentifierEnd}`,
+    'g'
+  );
+
+  let contentSinceLastIdentifier = '';
+  for await (const fragment of inputIterable) {
+    contentSinceLastIdentifier += fragment;
+
+    let lastIndex = 0;
+    let match;
+    while ((match = regex.exec(contentSinceLastIdentifier)) !== null) {
+      const endIndex = regex.lastIndex;
+
+      // Stream content up to the end of the identifier.
+      const contentToStream = contentSinceLastIdentifier.slice(
+        lastIndex,
+        endIndex
+      );
```

Comment:
Do you know if the code block can be mixed with the normal text withing one fragment? Or maybe the code block always takes the whole fragment.

Reply:
There are tests for both cases, code block all in one fragment, code block and identifiers split between fragments. Typically the fragments are 1-3 characters.